encryption/src/main/java/org/apache/solr/encryption/EncryptionRequestHandler.java

```diff
@@ -412,58 +409,46 @@ private void distributeRequest(SolrQueryRequest req, SolrQueryResponse rsp, Stri
-    params.set(PARAM_KEY_ID, keyId == null ? NO_KEY_ID : keyId);
-    return params;
-  }
-
-  boolean isTimeout(long maxTimeNs) {
-    return System.nanoTime() > maxTimeNs;
+    return new ModifiableSolrParams().set(PARAM_KEY_ID, keyId == null ? NO_KEY_ID : keyId);
   }
 
-  private State sendEncryptionRequestWithRetry(Replica replica, ModifiableSolrParams params, Http2SolrClient httpSolrClient, String keyId, String collection) {
+  private State sendEncryptionRequestWithRetry(
+      Replica replica,
+      String requestPath,
+      ModifiableSolrParams params,
+      Http2SolrClient httpSolrClient,
+      String keyId) {
     for (int numAttempts = 0; numAttempts < DISTRIBUTION_MAX_ATTEMPTS; numAttempts++) {
       try {
-        NamedList<Object> response = sendEncryptionRequest(replica, params, httpSolrClient);
-        Object responseStatus = response.get(STATUS);
-        Object responseState = response.get(ENCRYPTION_STATE);
-        log.info("Encryption state {} status {} for replica {} keyId {} collection={}", responseStatus, responseState, replica.getName(), keyId, collection);
+        SimpleSolrResponse response = sendEncryptionRequest(replica, requestPath, params, httpSolrClient);
+        Object responseStatus = response.getResponse().get(STATUS);
+        Object responseState = response.getResponse().get(ENCRYPTION_STATE);
+        log.info("Encryption state {} status {} for replica {} keyId {} in {} ms", responseStatus, responseState, replica.getName(), keyId, response.getElapsedTime());
         if (responseState != null) {
           return State.fromValue(responseState.toString());
         }
       } catch (SolrServerException | IOException e) {
         log.error("Error occurred while sending encryption request", e);
       }
     }
-    throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Failed encryption request to replica " + replica.getName() + " for keyId " + keyId + " collection " + collection);
+    throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Failed encryption request to replica " + replica.getName() + " for keyId " + keyId);
   }
 
-  private NamedList<Object> sendEncryptionRequest(Replica replica, ModifiableSolrParams params, Http2SolrClient httpSolrClient)
+  private SimpleSolrResponse sendEncryptionRequest(
+      Replica replica,
+      String requestPath,
+      ModifiableSolrParams params,
+      Http2SolrClient httpSolrClient)
       throws SolrServerException, IOException {
-    GenericSolrRequest request = new GenericSolrRequest(SolrRequest.METHOD.POST, getPath(), params);
-    request.setBasePath(replica.getCoreUrl());
-    return httpSolrClient.request(request);
-  }
-
-  protected String getPath() {
-    return "/admin/encrypt";
+    GenericSolrRequest distributedRequest = new GenericSolrRequest(SolrRequest.METHOD.POST, requestPath, params);
+    return httpSolrClient.requestWithBaseUrl(replica.getCoreUrl(), replica.getCollection(), distributedRequest);
```

Comment:
if you use getCoreUrl, it's complete (solr (aka "base url") + core).  Passing collection here would be a problem; concatenating that after the core url which doesn't make sense.  A test had better fail here.

Reply:
Thanks for the info. It's not obvious from the Http2SolrClient.requestWithBaseUrl doc that the collection name is concatenated.
When looking at the internals, it seems to me the concatenation happens in ClientUtils.buildRequestUrl, and it checks if solrRequest.requiresCollection(), which is false in the case here. That explains why tests work.
I'll pass null for the collection name to be explicit.

Reply:
CC @gerlowskija I propose `org.apache.solr.client.solrj.util.ClientUtils#buildRequestUrl` fail with `IllegalArgumentException` if `collection` is passed in but isn't used.